Main.hs

```diff
@@ -1,107 +1,95 @@
+{-# LANGUAGE MultiWayIf, BangPatterns #-}
+
 module Main where
 
 import Control.Monad
 import System.IO
+import qualified Data.List as List
 
-square :: Int -> Pic
-square n = Pic $ replicate n (replicate n '#')
+newtype Tixel = Tixel { unTixel :: Char }
 
-triangle :: Int -> Pic
-triangle = Pic . aux
-  where
-    aux :: Int -> [String]
-    aux 0 = []
-    aux n = replicate n 'T':aux (n-1)
+instance Semigroup Tixel where
+  tx1@(Tixel px1) <> tx2 = if px1 == ' ' then tx2 else tx1
 
-circle :: Int -> Pic
-circle r = Pic [ [ if x*x + y*y < r*r then 'S' else ' '
-                 | x <- [-r..r]
-                 ]
-               | y <- [-r..r]
-               ]
+instance Monoid Tixel where
+  mempty = Tixel ' '
 
-newtype Pic = Pic [String]
+newtype Pic a = Pic ((Int, Int) -> a)
 
-draw :: Pic -> IO ()
-draw (Pic p) = do
-  putStr "\ESC[2J"  -- clear screen
-  -- forM_ p putStrLn -- really slow
-  putStrLn $ unlines p -- also slow
-
--- Lines need to be padded on the right as well
--- This breaks the symmetry between the two functions
---  though: suspect it might be better done elsewhere
--- Right-padding is a waste of space: there isn't
---  anything to represent, except at the rendering
---  stage, where it can be done trivially (and flexibly)
-overlay :: Pic -> Pic -> Pic
-overlay (Pic p1) (Pic p2) = Pic $ zipWith combine pp1 pp2
-  where
-    maxHeight = max (length p1) (length p2)
-    pad n ss = ss <> replicate n " "
-    (pp1, pp2) = ( pad (maxHeight - length p1) p1
-                 , pad (maxHeight - length p2) p2)
-    combine = overlayLine
-
-overlayLine :: String -> String -> String
-overlayLine s1 s2 = zipWith combine p1 p2
+instance Semigroup a => Semigroup (Pic a) where
+  (Pic f1) <> (Pic f2) = Pic $ \p -> (f1 p) <> (f2 p)
+
+instance Monoid a => Monoid (Pic a) where
+  mempty = Pic $ \p -> mempty
+
+instance Functor Pic where
+  fmap f (Pic pf) = Pic $ f . pf
+
+draw :: (Int, Int) -> Pic Tixel -> String
+draw (w, h) (Pic pf) =
+  List.intercalate
+    ['\n']
+    [[unTixel $ pf (x, y) | x <- [-hw..hw]] | y <- [-hh..hh]]
   where
-    pad n s = s <> replicate n ' '
-    maxLen = max (length s1) (length s2)
-    (p1, p2) = ( pad (maxLen - length s1) s1
-               , pad (maxLen - length s2) s2)
-    combine u o = if o == ' ' then u else o
+    (hw, hh) = (w `div` 2, h `div` 2)
 
--- isn't there a Functorish thing for this?
-fm f (Pic p) = Pic (f p)
+square :: Monoid a => a -> Int -> Pic a
+square px n = Pic $ \(x, y) -> if abs x < n && abs y < n then px else mempty
 
-crop :: Int -> Int -> Pic -> Pic
-crop w h = fm (take h . map (take w))
+poly :: a -> [(Int, Int)] -> Pic a
```

Comment:
In general graphics, yes: in TUI this might be less useful than, say, a mesh or grid primitive (like for making mazes).

Reply:
I wanted to reimplement the triangle function but I was lazy to type out the three point arguments so I made it a list, but then it became a polygon.

I do agree that polygons don't have the same pop in 80*25 text.